### ISSUE TITLE
Allow downloads without a public URL

### ISSUE THREAD

**Issue by jcdarwin** [doublesecretagency/craft-digitaldownload#1]
At the moment, this plugin expects that you select `Assets in this volume have public URLs` and define a `Base URL` in the `Settings > Assets > Volumes` screen.

However, if we want to ensure that our assets are not accessible other than through the craft-digitaldownload plugin, we need to prevent public URLs.

Accordingly, `craft-digitaldownload/src/services/Download.php` should change from:

```
    private function _outputFile($link)
    {
        $asset = $link->asset();
        if ($asset) {
            header('Content-type: application/octet-stream');
            header('Content-disposition: attachment; filename='.$asset->filename);
            echo file_get_contents($asset->url);
            exit;
        }
    }

```

to something like:

```
    private function _outputFile($link)
    {
        $asset = $link->asset();
        if ($asset) {
            header('Content-type: application/octet-stream');
            header('Content-disposition: attachment; filename='.$asset->filename);

            // header("Content-Length: " . $asset->size);
            $assetFilePath = $this->_getAssetFilePath($asset);
            echo file_get_contents($assetFilePath);
            exit;
        }
    }

    // ========================================================================

    private function _getAssetFilePath($asset)
    {
        $volumePath = $asset->getVolume()->settings['path'];
        $folderPath = $asset->getFolder()->path . '/';
        $assetFilePath = Craft::getAlias($volumePath) . $folderPath . $asset->filename;

        return $assetFilePath;
    }
```

I.e. instead of using `$asset->url` with `file_get_contents`, we instead should use `$assetFilePath`.

As this change is quite small, I've not done this as a pull request, but let me know if you'd like me to.

**Comment by lindseydiloreto** [comment]
Good tip, thanks! I'll take a look into it and possibly ping you for a PR. 👍

**Comment by rob-c-baker** [comment]
Hi there,

I am running into this issue too.

We are attempting to prevent public access to some files that are not in a web accessible folder (i.e. no public URLs possible for the Craft volume), unless accessed by a use count controlled, expiring link.

I also have some pretty large files (>100Mb) that need to be delivered, so maybe [this is the way to do it](https://www.php.net/manual/en/function.readfile.php) instead of `file_get_contents()`?

I would like to use this plugin over LinkVault as it has the features we need!

Thanks.

**Comment by lindseydiloreto** [comment]
Awesome, thanks for submitting that! I'll take a look at your PR and (theoretically) push it through. 👍 

Out of curiosity, which features make you prefer this plugin over LinkVault?

**Comment by lindseydiloreto** [comment]
Looks great, thanks @rob-baker-ar (and @jcdarwin)! 🍺 

I've merged in these changes. I'm going to make a few additional updates before I tag a new release and push it live officially.

In the meantime, you can access these updates by doing this:

1. Edit `composer.json`
2. Change the plugin version to `v2.x-dev`
3. Run `composer update`

Thanks again!

**Comment by rob-c-baker** [comment]
@lindseydiloreto it is the fact that I can store a token that essentially defines the rules for download of an arbitrary file. That approach allows me to set a number of times a file can be downloaded, in my case, per order / customer (I'm using Commerce). With LinkVault it seems I could set a number of times a file can be downloaded but that seemed to apply globally to that one file (however many people purchase access to that file, everyone who accesses collectively gets counted as a download of that file).

Aside from that it would be very useful if I could store a field with the generated token (i.e. order id) so I can relate it to an order instead of having to store a token on the order itself... That is something LinkVault offers ;)

**Comment by lindseydiloreto** [comment]
Ok great, thanks for the feedback @rob-baker-ar!

Take a look at this technique, which should allow you to store a token directly on the order or customer record...

>https://www.doublesecretagency.com/plugins/digital-download/docs/storing-a-token

**Comment by rob-c-baker** [comment]
That is the exact functionality I am using.

Because any order we have may have multiple line items for digital products, we need a matrix on the order for storing the tokens (I have plugin code that generates the tokens and saves them in there when an order is paid). I was just thinking it would be nice to be able to store a reference to the order in the token (on the token db table?), a 2 way relationship between the order and token, so in the cases we will certainly get... it becomes a bit easier to give customers that mess things up a few more download attempts by adjusting the `totalDownloads` column.

..But yeah, for us, it's a nice-to-have rather than essential, I know we can get there by comparing the token with the stored ones in the matrix for the order. Complicates the SQL a bit, but meh.

**Comment by lindseydiloreto** [comment]
Good to know, thanks. I have long-term plans to add a GUI for the tokens in the CP, but I'm not sure it'll make it into the next release.

**Comment by rob-c-baker** [comment]
Any idea when the next release will be?

**Comment by lindseydiloreto** [comment]
Soon-ish... ideally this week.

I actually did a _major_ overhaul of the entire download mechanism yesterday. Apparently, even `readfile` wasn't covering all of our needs. With the new architecture in place, the plugin can now handle:

 - Local files without public URLs.
 - Cloud storage files (which broke with the `readfile` change)
 - Gigantic files! (ie: many gigabytes large)

If you've already got the plugin installed, you can switch the version to `v2.x-dev` in `composer.json`, then run `composer update`. That will grab these latest changes until I get a chance to push everything live. 👍

**Comment by lindseydiloreto** [comment]
This is officially released! Go ahead and grab **v2.1** whenever you get a chance.
